ocaml/xapi/vgpuops.ml

```diff
@@ -55,10 +55,46 @@ let fail_creation vm vgpu =
         Ref.string_of vgpu.gpu_group_ref
       ]))
 
-let allocate_vgpu_to_gpu ~__context vm host vgpu =
+let allocate_vgpu_to_gpu ?(dry_run=false) ?(pre_allocate_list=[]) ~__context vm host vgpu =
+  (* Get all pGPU from the host *)
   let available_pgpus = Db.Host.get_PGPUs ~__context ~self:host in
+  (* Get all pGPU from the required groups *)
   let compatible_pgpus = Db.GPU_group.get_PGPUs ~__context ~self:vgpu.gpu_group_ref in
   let pgpus = List.intersect compatible_pgpus available_pgpus in
+
+  let pgpu_can_hold_vgpu pgpu vgpu =
+    try Xapi_pgpu.assert_can_run_VGPU ~__context ~self:pgpu ~vgpu;true
+    with e -> false in
+  (* Get all pGPU that can hold the vGPU *)
+  let active_pgpus = List.filter ( fun pgpu -> pgpu_can_hold_vgpu pgpu vgpu.vgpu_ref ) pgpus in
+
+  let remaining_capacity_for_vgpu_from_pgpu vgpu pgpu =
+    let db_remaining =  Helpers.call_api_functions ~__context
+        (fun rpc session_id ->
+           Client.Client.PGPU.get_remaining_capacity ~rpc ~session_id
+             ~self:pgpu ~vgpu_type:vgpu.type_ref) in
+
+    let vGPU_size_of self other =
+      let size_of_vgpu vgpu_ref =
+        let vgpu_type = Db.VGPU.get_type ~__context ~self:vgpu_ref in
+        Db.VGPU_type.get_size ~__context ~self:vgpu_type in
+      let self_size = size_of_vgpu self in
+      let other_size = size_of_vgpu other in
+      (Int64.to_float self_size) /. (Int64.to_float other_size) in
+
+    (* Check if any pre_allocation existed, the pre_allocation is a set of vGPU allocation that
+     * not reflected in the database, usually in the dry run mode, with following format
+     * [(v1,p1);(v2,p2);(v3,p1)...]*)
+    let virtual_allocation = List.fold_left (fun num ele ->
+        match ele with
+        |(cvgpu,cpgpu) when cpgpu = pgpu -> num +. (vGPU_size_of cvgpu vgpu.vgpu_ref) 
```

Comment:
`vgpu.vgpu_ref` is constant over the list but its size is computed over and over again. I suspect `other` in `vGPU_size_of` is constant and could be computed once.

Reply:
good point, refined for `vgpu.vgpu_ref`,
I think `other` need to be queried from the database everytime, unless we `cache` the result in memory,
Does that complicated?